File: src/main/java/org/ivcode/beeboop/plugin/persistence/Datastore.java
package org.ivcode.beeboop.plugin.persistence;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Map;

public interface Datastore<V> {
    void put(@NotNull String key, @NotNull V value);

    V get(@NotNull String key);

    @NotNull
    List<String> listKeys();

    @NotNull
    Map<String, V> list();

    void delete(@NotNull String key);
}

File: src/main/kotlin/datastore/datastore/Datastore.kt
package org.ivcode.beeboop.datastore.datastore

import org.ivcode.beeboop.PROPERTY_PREFIX_APP_NAME

const val PROPERTY_NAME_DATASTORE_TYPE = "$PROPERTY_PREFIX_APP_NAME.datastore.type"

/**
 * Generic interface for a persistent key/value datastore
 */
interface Datastore<V> {
    fun put(key: String, value: V)
    fun get(key: String): V?
    fun listKeys(): List<String>
    fun list(): Map<String, V>
    fun delete(key: String)
}